vowpalwabbit/cb_explore_adf_greedy.cc

```diff
@@ -0,0 +1,117 @@
+#include "cb_explore_adf_greedy.h"
+#include "reductions.h"
+#include "cb_adf.h"
+#include "rand48.h"
+#include "bs.h"
+#include "gen_cs_example.h"
+#include "cb_explore.h"
+#include "explore.h"
+#include <vector>
+#include <algorithm>
+#include <cmath>
+#include <functional>
+
+namespace VW
+{
+namespace cb_explore_adf
+{
+namespace greedy
+{
+template <bool is_learn>
+void cb_explore_adf_greedy::predict_or_learn_impl(LEARNER::multi_learner& base, multi_ex& examples)
+{
+  // Explore uniform random an epsilon fraction of the time.
+  LEARNER::multiline_learn_or_predict<is_learn>(base, examples, examples[0]->ft_offset);
+
+  ACTION_SCORE::action_scores& preds = examples[0]->pred.a_s;
+
+  uint32_t num_actions = (uint32_t)preds.size();
+
+  size_t tied_actions = fill_tied(preds);
+
+  const float prob = m_epsilon / num_actions;
+  for (size_t i = 0; i < num_actions; i++) preds[i].score = prob;
+  if (!m_first_only)
+  {
+    for (size_t i = 0; i < tied_actions; ++i) preds[i].score += (1.f - m_epsilon) / tied_actions;
+  }
+  else
+    preds[0].score += 1.f - m_epsilon;
+}
+
+template <bool is_learn>
+void cb_explore_adf_greedy::predict_or_learn(
+    cb_explore_adf_greedy& data, LEARNER::multi_learner& base, multi_ex& examples)
+{
+  if (is_learn)
+    data.learn(data, &cb_explore_adf_greedy::predict_or_learn_impl<true>,
+        &cb_explore_adf_greedy::predict_or_learn_impl<false>, base, examples);
+  else
+    data.predict(data, &cb_explore_adf_greedy::predict_or_learn_impl<false>, base, examples);
+}
+
+void finish_multiline_example(vw& all, cb_explore_adf_greedy& data, multi_ex& ec_seq)
+{
+  data.finish_multiline_example(all, ec_seq);
+}
+
+void finish(cb_explore_adf_greedy& data) { data.~cb_explore_adf_greedy(); }
+
+LEARNER::base_learner* setup(VW::config::options_i& options, vw& all)
+{
+  using config::make_option;
+  using namespace std::placeholders;
+  auto data = scoped_calloc_or_throw<cb_explore_adf_greedy>();
```

Comment:
If you make epsilon and first_only as constructor args, and declare and use stack variables here, you don't have to construct an instance of cb_explore_adf_greedy.  Also then m_epsilon and m_first_only can be private.  Do you see any downside to that?

Reply:
Now that Jack's changes for scoped_calloc_or_throw (to allow argument forwarding to the constructor) are in, I can start hiding member variables and passing them in via constructor